%\VignetteEngine{knitr::knitr}
%\VignetteIndexEntry{Grade of Membership Clustering and Visualization using CountClust}
%\VignettePackage{CountClust}

% To compile this document
% library('knitr'); rm(list=ls()); knit('CountClust/vignettes/count-clust.Rnw')
% library('knitr'); rm(list=ls()); knit2pdf('CountClust/vignettes/count-clust.Rnw'); openPDF('count-clust.pdf')
% !Rnw weave = knitr

\documentclass[12pt]{article}

\newcommand{\ecostructure}{\textit{ecostructure}}
\usepackage{dsfont}
\usepackage{cite}


<<knitr, echo=FALSE, results="hide">>=
library("knitr")
opts_chunk$set(tidy=FALSE,tidy.opts=list(width.cutoff=30),dev="png",fig.show="hide",
               fig.width=4,fig.height=7,
               message=FALSE)
@

<<style, eval=TRUE, echo=FALSE, results="asis">>=
BiocStyle::latex()
@


\bioctitle[\ecostructure{} - Grade of Membership Model and Visualization for ecological species abundance data]{\ecostructure{} - Grade of Membership Model and Visualization for ecological species abundance data}

\begin{document}

\maketitle

\section{Introduction}

The \textbf{ecostructure} package is an R package that replicates the statistical analysis in this paper, but its toolbox of functions is generic enough in handling and analyzing other species abundance data. The package provides functions for fitting the Grade of Membership (GoM) model, along with the visualization of model fit using Block Structure plot \ref{Pritchard}\ref{Taddy}. The package comes with the raw taxonomic data saved as an ExpressionSet object and provides a pipeline for reading and processing counts data for different dimensions of diversity, e.g. - phylogenetic, regional and functional, which serve as readymade input for the GoM model. This package is an upgraded version of the CountClust package due to Dey et al \ref{Dey} for fitting GoM models on RNA-seq data

\section{Installation}

The package is available on Github and can be installed as follows

<<options, results="hide", echo=FALSE>>=
options(digits=3, width=80, prompt=" ", continue=" ")
@

<<install_ecostructure_bio, eval=FALSE>>=
library(devtools)
install_github("kkdey/ecostructure")
@ %def

Load the package as

<<load_ecostructure_bio, eval = TRUE>>=
library(ecostructure)
library(Biobase)
@ %def

to use the GoM model, the user needs to install the **maptpx** package

<<install_github, eval=FALSE>>=
library(devtools)
install_github("kkdey/maptpx")
@ %def


\section{Data Preparation}

One can load the taxonomic data, together with the grid metadata ans species metadata as an ExpressionSet object as follows

<<data, eval=TRUE>>=
data <- get(load(system.file("extdata", "HimalayanBirdsData.rda",
                             package = "ecostructure")))
taxonomic_counts <- exprs(data)
taxonomic_counts[1:5,1:5]
@ %def


The corresponding grid metadata can be read as

<<pdata, eval = TRUE>>=
grid_metadata <- pData(phenoData(data))
head(grid_metadata)
@ %def

The species metadata can be read as follows

<<fdata, eval = TRUE>>=
species_metadata <- pData(featureData(data))
head(species_metadata)
@ %def


Along with the taxonomic data and metadata, the package provides the phylogenetic tree data for the bird species as a **.tre** file that can be loaded as follows, using the package **ape**.

<<tree, eval = TRUE>>=
phylo_tree <- ape::read.tree(system.file("extdata", "AllHim_Mar_27_2015.tre",
                             package = "ecostructure"))
phylo_tree
@ %def


The shape files for the regional motif analysis can be loaded as follows

<<shp, eval = TRUE>>=
shp_file <- ape::read.tree(system.file("extdata", "AllHim_Mar_27_2015.tre",
                             package = "ecostructure"))
shp_file
@ %def

\section{Grade of Membership Model and Visualization}

Here we illustrate how one can fit the Grade of Membership model and perform the visualization of the model fit using the Block Structure plot. Here we present a case study with number of clusters chosen to be between $2$ and $4$.

<<blockstructure, eval=TRUE>>=
elevation_metadata=grid_metadata$Elevation;
east_west_dir = grid_metadata$WorE;
gom_fit <- CountClust::FitGoM(t(taxonomic_counts), K=2:4, tol=0.1)
@ %def


\begin{verb} gom_fit \end{verb} is a list of size 3, with each component representing the model fit for the cluster $k$, varying from 2 to 4. The two main components of the model fit are the membership proportion matrix $\omega$, given by \begin{verb} gom_fit[[k]]$omega \end{verb} and the motif matrix \begin{verb} gom_fit[[k]]$theta \end{verb}. Examples for $K=2$ are

<<omega, eval=TRUE>>=

omega <- gom_fit[[2]]$omega
head(omega)
rowSums(omega)


theta <- gom_fit[[2]]$theta
head(theta)
colSums(theta)

@ %def

Using the grid metadata, we provide a Block Structure Plot representation of the membership proportion matrix. In a block Structure plot representation, one metadata is used for forming blocks (here the East/West direction) and in each block, the
the samples (along the rows of the Structure plot) are arranged by a second metadata (say Elevation).

<<block_structure, eval=TRUE, fig.height = 4, fig.height = 6>>=
BlockStructure(omega, blocker_metadata = east_west_dir,
               order_metadata = elevation_metadata,
               yaxis_label = "Elevation",
               levels_decreasing = FALSE)
@ %def


\begin{figure}[htp]
\begin{center}
\includegraphics[width=6in,height=6in]{figure/block_structure.png}
\end{center}
\end{figure}

\newpage

\section{Processing motif data}

\section{Extras}

\textbf{ecostructure} provides additional functions for plotting a variable of interest (which could be the grades of membership or the motif pattern) against a metadata and a diversity measure or against two metadata in three way scatter plot functions.

an example of plotting a variable against two metadata are as follows.

<<topic_meta_meta, eval=TRUE, fig.height = 4, fig.height = 6>>=
annotation = data.frame(x_names = c(paste0("A",1:5), paste0("B",1:5)),
x = c(0.5,2.0, 3.2, 4.6, 6.3,  23.5, 26.4, 28.5, 29.6, 31.8),
y1 = c(0.4, 0.3, 0.4, 0.35, 0.4, 0.8, 0.85, 0.9, 0.8, 0.75),
y2 =c(5, 6.6, 4, 5.2, 20, 3.4, 5.6, 4.5, 8, 10))

head(annotation)

topic_meta_meta(annotation)
@

An example of plotting  a variable against two diversity measures as follows.

<<topic_meta_diversity, eval=TRUE, fig.height = 4, fig.height = 6>>=
annotation = data.frame(x_names = c(paste0("A",1:5), paste0("B",1:5)),
    x = c(0.5,2.0, 3.2, 4.6, 6.3,  23.5, 26.4, 28.5, 29.6, 31.8),
    y1 = c(0.4, 0.3, 0.4, 0.35, 0.4, 0.8, 0.85, 0.9, 0.8, 0.75),
    y2d =c(5, 6.6, 4, 5.2, 20, 3.4, 5.6, 4.5, 8, 10))

head(annotation)

topic_meta_diversity(annotation)

@


\end{document}
